<a href="https://colab.research.google.com/github/Gulzaibsharif-coder/flyrank-ml-capstone/blob/main/work/notebooks/w03_data_contract.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-04 — Search Intelligence Data Contract

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Gulzaibsharif-coder/flyrank-ml-capstone/blob/main/work/notebooks/w03_data_contract.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. Unit of analysis + time window

*One row = one what, over which dates? State it, then verify it below.*

### Unit of analysis
One row represents one search performance record for a page (URL) within a specific reporting date or time window.

### Time window
The analysis uses the reporting period available in the FlyRank warehouse dataset.

In [12]:
!pip -q install datasets duckdb pandas pyarrow

In [14]:
from google.colab import userdata
from datasets import load_dataset

HF_TOKEN = userdata.get("HF_TOKEN")

dataset = load_dataset(
    "FlyRank/internship-warehouse",
    "fact_content_daily_performance",
    token=HF_TOKEN
)

print(dataset)

Resolving data files:   0%|          | 0/18 [00:00<?, ?it/s]

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 1.45MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B /  624kB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 19.6kB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 4.41MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 3.29MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 3.22MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 2.62MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 72.0MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 89.0MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 8.93MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 86.2MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 90.3MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B /  134MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B /  124MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 7.12MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B / 21.6MB            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B /  149MB            

fact_content_daily_performance/month=202(…): reconstructing file:   0%|          |  0.00B /  146MB            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

fact_content_daily_performance/month=202(…): downloading bytes:           |  0.00B            

Generating train split:   0%|          | 0/78835655 [00:00<?, ? examples/s]

Loading dataset shards:   0%|          | 0/39 [00:00<?, ?it/s]

DatasetDict({
    train: Dataset({
        features: ['report_date', 'client_hash_id', 'content_hash_id', 'client_has_gsc', 'client_has_ga4', 'gsc_data_available', 'ga4_data_available', 'gsc_impressions', 'gsc_clicks', 'gsc_sum_position', 'gsc_avg_position', 'ga4_pageviews', 'ga4_sessions', 'ga4_users', 'ga4_engaged_sessions', 'ga4_total_engagement_sec', 'sessions_organic', 'sessions_direct', 'sessions_referral', 'sessions_social', 'sessions_paid', 'sessions_ai', 'ai_chatgpt', 'ai_perplexity', 'ai_gemini', 'ai_copilot', 'ai_claude', 'ai_meta', 'ai_other', 'scroll_events'],
        num_rows: 78835655
    })
})


In [15]:
print(dataset["train"].column_names)

['report_date', 'client_hash_id', 'content_hash_id', 'client_has_gsc', 'client_has_ga4', 'gsc_data_available', 'ga4_data_available', 'gsc_impressions', 'gsc_clicks', 'gsc_sum_position', 'gsc_avg_position', 'ga4_pageviews', 'ga4_sessions', 'ga4_users', 'ga4_engaged_sessions', 'ga4_total_engagement_sec', 'sessions_organic', 'sessions_direct', 'sessions_referral', 'sessions_social', 'sessions_paid', 'sessions_ai', 'ai_chatgpt', 'ai_perplexity', 'ai_gemini', 'ai_copilot', 'ai_claude', 'ai_meta', 'ai_other', 'scroll_events']


## 2. Fields: feature / label / context / excluded

*Sort every field you plan to touch into these four buckets. Excluded needs a why.*

### Features
- report_date
- content_hash_id
- gsc_impressions
- gsc_clicks
- gsc_ctr
- gsc_position
- ga4_users
- ga4_sessions
- ga4_engagement_rate
- ga4_conversions

### Label
- content_refresh_needed (derived from search performance trends)

### Context
- client_hash_id
- client_has_gsc
- client_has_ga4
- gsc_data_available
- ga4_data_available

### Excluded
- Anonymous identifiers not required for prediction beyond grouping.
- Fields not related to search performance or content refresh decisions.

In [16]:
for col in dataset["train"].column_names:
    print(col)


report_date
client_hash_id
content_hash_id
client_has_gsc
client_has_ga4
gsc_data_available
ga4_data_available
gsc_impressions
gsc_clicks
gsc_sum_position
gsc_avg_position
ga4_pageviews
ga4_sessions
ga4_users
ga4_engaged_sessions
ga4_total_engagement_sec
sessions_organic
sessions_direct
sessions_referral
sessions_social
sessions_paid
sessions_ai
ai_chatgpt
ai_perplexity
ai_gemini
ai_copilot
ai_claude
ai_meta
ai_other
scroll_events


## 3. Verify it with queries (grain, counts, missing values, windows)

*Every claim above gets a query cell here. A contract claim without a query next to it is a guess.*

### Verification

The data contract was verified using dataset queries.

- Total records: 78,835,655
- Sample records were successfully retrieved.
- Each row contains a reporting date, client identifier, content identifier, and associated Google Search Console (GSC) and Google Analytics 4 (GA4) metrics.
- The dataset grain is one row per content page for a reporting date.

In [17]:
dataset["train"].num_rows


78835655

In [18]:
dataset["train"][:5]

{'report_date': [datetime.date(2025, 1, 27),
  datetime.date(2025, 1, 27),
  datetime.date(2025, 1, 27),
  datetime.date(2025, 1, 27),
  datetime.date(2025, 1, 27)],
 'client_hash_id': ['client_9958f0a7ae1df715',
  'client_9958f0a7ae1df715',
  'client_9958f0a7ae1df715',
  'client_9958f0a7ae1df715',
  'client_9958f0a7ae1df715'],
 'content_hash_id': ['content_3b70a18ea133b2bb',
  'content_fe8e8155ce1d47a2',
  'content_b4462a1b90640058',
  'content_c899aef92518c714',
  'content_c7c1d2e68d9d0964'],
 'client_has_gsc': [True, True, True, True, True],
 'client_has_ga4': [True, True, True, True, True],
 'gsc_data_available': [True, True, True, True, True],
 'ga4_data_available': [False, False, False, False, False],
 'gsc_impressions': [30, 5, 1, 6, 5],
 'gsc_clicks': [0, 0, 0, 0, 0],
 'gsc_sum_position': [115, 358, 34, 140, 89],
 'gsc_avg_position': [3.8333333333333335,
  71.6,
  34.0,
  23.333333333333332,
  17.8],
 'ga4_pageviews': [0, 0, 0, 0, 0],
 'ga4_sessions': [0, 0, 0, 0, 0],
 'ga4_use

In [19]:
dataset["train"][0]

{'report_date': datetime.date(2025, 1, 27),
 'client_hash_id': 'client_9958f0a7ae1df715',
 'content_hash_id': 'content_3b70a18ea133b2bb',
 'client_has_gsc': True,
 'client_has_ga4': True,
 'gsc_data_available': True,
 'ga4_data_available': False,
 'gsc_impressions': 30,
 'gsc_clicks': 0,
 'gsc_sum_position': 115,
 'gsc_avg_position': 3.8333333333333335,
 'ga4_pageviews': 0,
 'ga4_sessions': 0,
 'ga4_users': 0,
 'ga4_engaged_sessions': 0,
 'ga4_total_engagement_sec': 0,
 'sessions_organic': 0,
 'sessions_direct': 0,
 'sessions_referral': 0,
 'sessions_social': 0,
 'sessions_paid': 0,
 'sessions_ai': 0,
 'ai_chatgpt': 0,
 'ai_perplexity': 0,
 'ai_gemini': 0,
 'ai_copilot': 0,
 'ai_claude': 0,
 'ai_meta': 0,
 'ai_other': 0,
 'scroll_events': 0}

In [20]:
example = dataset["train"][:1000]

for col, values in example.items():
    missing = sum(v is None for v in values)
    print(f"{col}: {missing}")

report_date: 0
client_hash_id: 0
content_hash_id: 0
client_has_gsc: 0
client_has_ga4: 0
gsc_data_available: 0
ga4_data_available: 0
gsc_impressions: 0
gsc_clicks: 0
gsc_sum_position: 0
gsc_avg_position: 0
ga4_pageviews: 0
ga4_sessions: 0
ga4_users: 0
ga4_engaged_sessions: 0
ga4_total_engagement_sec: 0
sessions_organic: 0
sessions_direct: 0
sessions_referral: 0
sessions_social: 0
sessions_paid: 0
sessions_ai: 0
ai_chatgpt: 0
ai_perplexity: 0
ai_gemini: 0
ai_copilot: 0
ai_claude: 0
ai_meta: 0
ai_other: 0
scroll_events: 0


## 4. Data limits

*What can this data never tell you? Unbalanced history, GSC-only early rows, window overlaps.*

### Data Limits

- The dataset contains anonymized client and content identifiers.
- The analysis is limited to the reporting period available in the FlyRank warehouse.
- Some performance metrics depend on Google Search Console (GSC) and Google Analytics 4 (GA4) availability.
- Results should not be generalized beyond the available observation period.

In [22]:
sample = dataset["train"][:10000]

dates = sample["report_date"]

print("Earliest report date in sample:", min(dates))
print("Latest report date in sample:", max(dates))
print("Unique reporting dates in sample:", len(set(dates)))


Earliest report date in sample: 2025-01-27
Latest report date in sample: 2025-02-14
Unique reporting dates in sample: 10


## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.